html/janus.js

```diff
@@ -2196,6 +2196,10 @@ function Janus(gatewayCallbacks) {
 				if(media.addData) {
 					media.data = true;
 				}
+				// Reset external stream flag when replacing MediaStream with user media
+				if(config.streamExternal) {
+					config.streamExternal = false;
+				}
```

Comment:
But doesn't this mean that the user loses control of the `MediaStream` they provided? Resetting this flag means we'll close the stream when we're done, which the user may not want to happen if they did the `getUserMedia` themselves and just gave us the new stream to use.

Reply:
To get to this part of the code, the following conditionals have already been met:
Line 2070 - `!config.pc` is falsey. There is already a PeerConnection.
Line 2080 - `callbacks.stream` is falsey. There isn't an external stream present.

If `config.streamExternal === true` and there isn't an external stream, we'll want to set this flag `false` to clean up.

This flag reset is for the case where the application sends another `createOffer` without an external `stream`, after an initial `createOffer` with an external `stream` _was_ sent. Example: see the second code block in my initial PR description.


I did some testing with this `config.streamExternal = false;` line removed and the renegotiation does work, but it probably should be reset.

I did some tests in my Janus VideoRoom plugin application. We start all participants as publishers with a standard getUserMedia for video and audio and subscribe to any other publisher upon entering the room. One of the participants can use a canvas to replace their video stream. I did some tests where I:

1. Made a `createOffer` with a new external `stream` and `replaceVideo: true`, replacing the userVideo device.
2. Made a `createOffer` with a new external `stream` and `replaceVideo: true`, replacing the previous external `stream`.
3. Made a `createOffer` with user media and `replaceVideo: true`, replacing the external `stream`.

In test case 3, `config.streamExternal === true` at line 2200. It doesn't effect the rest of the flow of the `prepareWebrtc` function, but it seems like it should be cleaned up if it is `true` at that point.

Reply:
Thanks for the explanation, but I still have some doubts. When people provide an external stream, we never mess with it, as the application may be using it for other things as well at the same time, and so it's important we don't remove its tracks for instance, as otherwise any usage the developer may be doing of the same MediaStream somewhere else may be broken as a result. I guess it's probably hard to enforce that, especially with renegotiations, so if you can confirm this patch does the job in the cases you tested it in, it's good enough for me, and we can maybe loosen the assumptions for the future (meaning that if what I highlighted is indeed an issue, we can make this the developer's responsibility to give us a stream we can mess with, if needed).

Reply:
What I can confirm in my tests is that setting `config.streamExternal = false;` here does not have an effect on what I'm trying to achieve. I do not want this PR to break backwards compatibility, especially if I don't have a solid method of testing it, so I will remove it.